src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+                children
+                    .get_mut(*child)
+                    .unwrap()
+                    .offload_buffer(needed_memory)
+                    .await
+            }
+        }
+    }
+
+    fn merge_filters(dest: &mut Option<Bloom>, source: &Option<Bloom>) {
+        match source {
+            Some(source_filter) => {
+                let res = match dest {
+                    Some(dest_filter) => dest_filter.checked_add_assign(source_filter),
+                    _ => None,
+                };
+                if res.is_none() {
+                    *dest = None;
+                }
+            }
+            None => *dest = None,
+        }
+    }
+
+    fn push_group(&mut self, value: Self) {
+        match self {
+            Self::Vec(filter, values) => {
+                match &value {
+                    Self::Vec(child_filter, _) => {
+                        Self::merge_filters(filter, child_filter);
+                    }
+                    _ => unreachable!(),
+                }
+                values.push(value);
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn push_item(&mut self, id: usize, child_filter: &Option<Bloom>) {
+        match self {
+            Self::Vec(filter, values) => {
+                values.push(Self::Item(id));
+                Self::merge_filters(filter, child_filter)
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn push(&mut self, id: usize, filter: &Option<Bloom>, group_size: usize) {
+        match self {
+            Self::Vec(_, values) => {
+                if values.is_empty() {
+                    values.push(Default::default())
+                }
+                let last = values.last_mut().expect("Have values");
+                last.push_item(id, filter);
+                if last.len() >= group_size {
+                    drop(last);
+                    let last = values.pop().expect("Have values");
+                    if values.is_empty() {
+                        values.push(Default::default())
+                    }
+                    let pred_last = values.last_mut().expect("Have values");
+                    pred_last.push_group(last);
+                    values.push(Default::default());
+                }
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    fn len(&self) -> usize {
+        match self {
+            Self::Vec(_, values) => values.len(),
+            _ => 0,
+        }
+    }
+
+    fn add_to_intermediate_filters(&mut self, id: usize, item: &[u8]) -> Result<bool> {
+        match self {
+            Self::Vec(filter, values) => {
+                if let Some(filter) = filter {
+                    let mut should_add = false;
+                    for value in values.iter_mut() {
+                        should_add = should_add || value.add_to_intermediate_filters(id, item)?;
+                        if should_add {
+                            filter.add(item)?;
```

Comment:
If you have a guarantee that only one item has specified `id` (ids are unique), then you should stop iterating over `values` here